bigquery/google/cloud/bigquery/table.py

```diff
@@ -616,41 +619,58 @@ def _set_properties(self, api_response):
+        resource['expirationTime'] = value
+
+    def _populate_partitioning_type_resource(self, resource):
+        resource['timePartitioning'] = self._properties['timePartitioning']
+
+    def _populate_view_use_legacy_sql_resource(self, resource):
+        if 'view' not in resource:
+            resource['view'] = {}
+        resource['view']['useLegacySql'] = self.view_use_legacy_sql
+
+    def _populate_view_query_resource(self, resource):
+        if 'view' not in resource:
+            resource['view'] = {}
+        resource['view']['query'] = self.view_query
+
+    def _populate_schema_resource(self, resource):
+        resource['schema'] = {'fields': _build_schema_resource(self._schema)}
+
+    def _build_resource(self, filter_fields=[]):
         """Generate a resource for ``create`` or ``update``."""
         resource = {
             'tableReference': {
                 'projectId': self._project,
                 'datasetId': self._dataset_id,
                 'tableId': self.table_id},
         }
-        if self.description is not None:
-            resource['description'] = self.description
-
-        if self.expires is not None:
-            value = _millis_from_datetime(self.expires)
-            resource['expirationTime'] = value
-
-        if self.friendly_name is not None:
-            resource['friendlyName'] = self.friendly_name
-
-        if self.location is not None:
-            resource['location'] = self.location
-
-        if self.partitioning_type is not None:
-            resource['timePartitioning'] = self._properties['timePartitioning']
-
-        if self.view_query is not None:
-            view = resource['view'] = {}
-            view['query'] = self.view_query
-            if self.view_use_legacy_sql is not None:
-                view['useLegacySql'] = self.view_use_legacy_sql
-
-        if self._schema:
-            resource['schema'] = {
-                'fields': _build_schema_resource(self._schema)
-            }
-
+        custom_resource_fields = {
+            'expires': self._populate_expires_resource,
+            'partitioning_type': self._populate_partitioning_type_resource,
+            'view_query': self._populate_view_query_resource,
+            'view_use_legacy_sql': self._populate_view_use_legacy_sql_resource,
+            'schema': self._populate_schema_resource
+        }
+        all_fields = [
+            'description', 'friendly_name', 'expires', 'location',
+            'partitioning_type', 'view_use_legacy_sql', 'view_query', 'schema'
+        ]
+        for f in all_fields:
+            if filter_fields and f not in filter_fields:
+                continue
+            if getattr(self, f) in (None, []) and f not in filter_fields:
+                continue
+            if f in custom_resource_fields:
+                custom_resource_fields[f](resource)
+            else:
+                # snake case to camel case
+                words = f.split('_')
```

Comment:
Add a TODO to factor out this code so both this and dataset can use it.

Reply:
Added TODO in https://github.com/GoogleCloudPlatform/google-cloud-python/pull/4076/commits/19af93abcde93af803b07f6f3ebe468e30c01c11